java/src/org/openqa/selenium/remote/http/RetryRequest.java

```diff
@@ -17,89 +17,55 @@
 
 package org.openqa.selenium.remote.http;
 
-import static java.net.HttpURLConnection.HTTP_CLIENT_TIMEOUT;
 import static java.net.HttpURLConnection.HTTP_INTERNAL_ERROR;
 import static java.net.HttpURLConnection.HTTP_UNAVAILABLE;
-import static org.openqa.selenium.internal.Debug.getDebugLogLevel;
-import static org.openqa.selenium.remote.http.Contents.asJson;
 
-import dev.failsafe.Failsafe;
-import dev.failsafe.Fallback;
-import dev.failsafe.RetryPolicy;
-import dev.failsafe.event.ExecutionAttemptedEvent;
-import dev.failsafe.function.CheckedFunction;
 import java.net.ConnectException;
-import java.util.Map;
+import java.util.logging.Level;
 import java.util.logging.Logger;
+import org.openqa.selenium.internal.Debug;
 
 public class RetryRequest implements Filter {
 
   private static final Logger LOG = Logger.getLogger(RetryRequest.class.getName());
+  private static final Level LOG_LEVEL = Debug.getDebugLogLevel();
 
-  private static final Fallback<HttpResponse> fallback =
-      Fallback.of(
-          (CheckedFunction<ExecutionAttemptedEvent<? extends HttpResponse>, ? extends HttpResponse>)
-              RetryRequest::getFallback);
-
-  // Retry on connection error.
-  private static final RetryPolicy<HttpResponse> connectionFailurePolicy =
-      RetryPolicy.<HttpResponse>builder()
-          .handleIf(failure -> failure.getCause() instanceof ConnectException)
-          .withMaxRetries(3)
-          .onRetry(
-              e ->
-                  LOG.log(
-                      getDebugLogLevel(),
-                      "Connection failure #{0}. Retrying.",
-                      e.getAttemptCount()))
-          .build();
-
-  // Retry if server is unavailable or an internal server error occurs without response body.
-  private static final RetryPolicy<HttpResponse> serverErrorPolicy =
-      RetryPolicy.<HttpResponse>builder()
-          .handleResultIf(
-              response ->
-                  response.getStatus() == HTTP_INTERNAL_ERROR
-                      && Integer.parseInt((response).getHeader(HttpHeader.ContentLength.getName()))
-                          == 0)
-          .handleResultIf(response -> (response).getStatus() == HTTP_UNAVAILABLE)
-          .withMaxRetries(2)
-          .onRetry(
-              e ->
-                  LOG.log(
-                      getDebugLogLevel(),
-                      "Failure due to server error #{0}. Retrying.",
-                      e.getAttemptCount()))
-          .build();
+  private static final int RETRIES_ON_CONNECTION_FAILURE = 3;
+  private static final int RETRIES_ON_SERVER_ERROR = 2;
+  private static final int NEEDED_ATTEMPTS =
+      Math.max(RETRIES_ON_CONNECTION_FAILURE, RETRIES_ON_SERVER_ERROR) + 1;
 
   @Override
   public HttpHandler apply(HttpHandler next) {
-    return req ->
-        Failsafe.with(fallback)
-            .compose(serverErrorPolicy)
-            .compose(connectionFailurePolicy)
-            .get(() -> next.execute(req));
-  }
+    return req -> {
```

Comment:
Adding some code comments along the flow will help others understand what is being done, as the logic behind the code is not trivial.